crates/wasi-common/yanix/src/sys/bsd/file.rs

```diff
@@ -8,15 +10,11 @@ pub unsafe fn isatty(fd: RawFd) -> Result<bool> {
         Ok(true)
     } else {
         // ... otherwise 0 is returned, and errno is set to indicate the error.
-        let errno = io::Error::last_os_error();
-        if let Some(raw_errno) = errno.raw_os_error() {
-            if raw_errno == libc::ENOTTY {
-                Ok(false)
-            } else {
-                Err(errno.into())
-            }
+        let errno = Error::last_os_error();
+        if errno.raw_os_error().unwrap() == libc::ENOTTY {
```

Comment:
I've seen this pattern in a few places at this point, and the `.unwrap()` here makes me a little uncomfortable. Could this perhaps be replaced with either a function that returns the last raw os error, or a function that extracts the raw os error and returns some other error if it can't be extracted? For example something like:

```rust
fn last_raw_os_error() -> i32 {
    Error::last_os_error().raw_os_error().unwrap_or(-1)
}
```

or something like:

```rust
fn raw_os_error(err: &io::Error) -> Result<i32> {
    err.raw_os_error().ok_or_else(Errno::EIO)
}
```

Reply:
The [docs for `raw_os_error`](https://doc.rust-lang.org/std/io/struct.Error.html#method.raw_os_error) guarantee that it'll return `Some` on an error returned from `last_os_error`, so the `unwrap` is guaranteed to succeed.

It doesn't matter so much in `isatty` here, but across wasi-common we often have subtle reasoning about specific error codes, so introducing new conditions under which errors like `EIO` or even `-1` could materialize makes that tricky.

Reply:
As @sunfishcode pointed out, according to the docs, this cannot ever fail as we only ever call `unwrap` after `io::Error` was populated with a raw OS error, either with `io::Error::last_os_error()` or `io::Error::from_raw_os_error()`. I do agree with you @alexcrichton that this is pretty subtle in the sense that we're relying on this prior knowledge that we've actually done just that. If it wasn't the case, we'd in a lot of trouble. Then again, since `yanix`, and in general, the syscall impl code in `wasi-common` _only_ (at least for now!) deals with raw OS errno values, this is safe. And since we're constantly evolving WASI, having panics around those calls might be a good thing to catch bugs that otherwise would be silently swallowed as @sunfishcode pointed out.